<launch>

    <node pkg="imu_filter_madgwick" type="imu_filter_node" name="imu_filter" output="screen">
        <param name="use_mag" type="bool" value="false"/>
        <param name="publish_tf" type="bool" value="false"/>
        <param name="world_frame" type="string" value="enu"/>
        <remap from="imu/data_raw" to="/camera/imu"/>
        <remap from="imu/data" to="/rtabmap/imu"/>
    </node>

<!-- Choose visualization -->
    <arg name="rviz" default="false" />
    <arg name="rtabmap_viz" default="true" />
    
    <param name="use_sim_time" type="bool" value="True"/>


<!-- Stereo Odometry -->   
    <node pkg="rtabmap_odom" type="stereo_odometry" name="stereo_odometry" output="screen">
        <remap from="left/image_rect"       to="/camera/infra1/image_rect_raw"/>
        <remap from="right/image_rect"      to="/camera/infra2/image_rect_raw"/>
        <remap from="left/camera_info"      to="/camera/infra1/camera_info"/>
        <remap from="right/camera_info"     to="/camera/infra2/camera_info"/>
        <remap from="odom"                  to="/stereo_odometry"/>

        <param name="frame_id" type="string" value="base_link"/>
        <param name="odom_frame_id" type="string" value="odom"/>
        <param name="imu_topic" type="string" value="/rtabmap/imu"/>
        <param name="wait_imu_to_init" type="bool" value="false"/>

        <param name="subscribe_rgbd" type="bool" value="false"/>
        <remap from="rgbd_image" to="/camera/color/image_raw"/>

        <param name="Odom/Strategy" type="string" value="0"/> <!-- 0=BOW, 1=OpticalFlow -->
        <param name="Odom/EstimationType" type="string" value="1"/> <!-- 3D->2D (PnP) -->
        <param name="Odom/MinInliers" type="string" value="10"/>
        <param name="Odom/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
        <param name="Odom/MaxDepth" type="string" value="10"/>
        <param name="OdomBow/NNDR" type="string" value="0.8"/>
        <param name="Odom/MaxFeatures" type="string" value="1000"/>
        <param name="Odom/FillInfoData" type="string" value="$(arg rtabmap_viz)"/>
        <param name="GFTT/MinDistance" type="string" value="10"/>
        <param name="GFTT/QualityLevel" type="string" value="0.00001"/> 
    </node>
<!-- Stereo SLAM -->
    <group ns="rtabmap">   
        <!-- Visual SLAM: args: "delete_db_on_start" and "udebug" -->
        <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="--delete_db_on_start">
            <param name="frame_id" type="string" value="base_link"/>
            <param name="subscribe_stereo" type="bool" value="true"/>
            <param name="subscribe_depth" type="bool" value="false"/> <!-- Can't be true at the same time as subscribe_stereo -->
            <!-- Make them true when connected to the LiDAR -->
            <param name="subscribe_scan" type="bool" value="false"/>
            <param name="subscribe_scan_cloud" type="bool" value="false"/>


            <!-- subscribe_stereo is set to true: -->
            <remap from="left/image_rect"       to="/camera/infra1/image_rect_raw"/>
            <remap from="right/image_rect"      to="/camera/infra2/image_rect_raw"/>
            <remap from="left/camera_info"      to="/camera/infra1/camera_info"/>
            <remap from="right/camera_info"     to="/camera/infra2/camera_info"/>

            <!-- subscribe_depth is set to true: -->
            <remap from="depth/image" to="/camera/depth/image_rect_raw"/>
            
            <!-- subscribe_scan is set to true: -->
            <remap from="scan" to="/scan"/>

            <!-- subscribe_scan_cloud is set to true: -->
            <remap from="scan_cloud" to="/velodyne_points"/>


            <remap from="odom" to="/stereo_odometry"/>

            <param name="queue_size" type="int" value="30"/>

            <!-- RTAB-Map's parameters -->
            <param name="imu_topic" type="string" value="/rtabmap/imu"/>
            <param name="Rtabmap/TimeThr" type="string" value="700"/>
            <param name="Rtabmap/DetectionRate" type="string" value="1"/>
            
            <param name="Kp/WordsPerImage" type="string" value="200"/>
            <param name="Kp/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
            <param name="Kp/DetectorStrategy" type="string" value="0"/>   <!-- use SURF -->
            <param name="Kp/NNStrategy" type="string" value="1"/>         <!-- kdTree -->

            <param name="SURF/HessianThreshold" type="string" value="1000"/>

            <param name="LccBow/MinInliers" type="string" value="10"/>
            <param name="LccBow/EstimationType" type="string" value="1"/> <!-- 3D->2D (PnP) -->

            <param name="LccReextract/Activated" type="string" value="true"/>
            <param name="LccReextract/MaxWords" type="string" value="500"/>
            <param name="LccReextract/MaxDepth" type="string" value="10"/> 
        </node>
        
        <!-- Visualisation RTAB-Map -->
        <node if="$(arg rtabmap_viz)" pkg="rtabmap_viz" type="rtabmap_viz" name="rtabmap_viz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
            <param name="subscribe_stereo" type="bool" value="true"/>
            <param name="subscribe_odom_info" type="bool" value="true"/>
            <param name="queue_size" type="int" value="10"/>
            <param name="frame_id" type="string" value="base_link"/>
            <remap from="left/image_rect"       to="/camera/infra1/image_rect_raw"/>
            <remap from="right/image_rect"      to="/camera/infra2/image_rect_raw"/>
            <remap from="left/camera_info"      to="/camera/infra1/camera_info"/>
            <remap from="right/camera_info"     to="/camera/infra2/camera_info"/>
            <remap from="odom_info" to="/odom_info"/>
            <remap from="odom" to="/stereo_odometry"/>
            <remap from="mapData" to="mapData"/>
        </node>
            
    </group>
    
    <!-- Visualisation RVIZ --> 
    <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap_demos)/launch/config/demo_stereo_outdoor.rviz"/>

</launch>